templates/root/spark/conf/spark-env.sh

```diff
@@ -14,8 +14,12 @@ export MASTER=`cat /root/spark-ec2/cluster-url`
 export SPARK_SUBMIT_LIBRARY_PATH="$SPARK_SUBMIT_LIBRARY_PATH:/root/ephemeral-hdfs/lib/native/"
 export SPARK_SUBMIT_CLASSPATH="$SPARK_CLASSPATH:$SPARK_SUBMIT_CLASSPATH:/root/ephemeral-hdfs/conf"
 
-# Bind Spark's web UIs to this machine's public EC2 hostname:
-export SPARK_PUBLIC_DNS=`wget -q -O - http://169.254.169.254/latest/meta-data/public-hostname`
+# Bind Spark's web UIs to this machine's public EC2 hostname otherwise fallback to private IP:
+SPARK_PUBLIC_DNS=`wget -q -O - http://169.254.169.254/latest/meta-data/public-hostname`
+if [[ -z "$SPARK_PUBLIC_DNS" ]]; then
+  SPARK_PUBLIC_DNS=`wget -q -O - http://169.254.169.254/latest/meta-data/local-ipv4`
```

Comment:
Glad you asked, 'cause I had some reasons:
 1. The behavior is consistent with the new "--private-ips" option for the spark_ec.py script in spark repo.
 1. Robust private hostname resolution is dependent on settings outside the purview of this script (e.g. VPN client and server settings) .  This env var is used to generate the links in the WebUI.  If we use hostnames and they don't resolve the links are essentially broken.  Host names are meant to be human friendly and there's no benefit to human friendliness in this context as the links are machine parsed.